contracts/staking/EpochSelector.sol

```diff
@@ -0,0 +1,1683 @@
+        uint256 sumOfLeftBalances,
+        uint256 balance,
+        uint256 sumOfRightBalances
+    )
+        internal
+        pure
+        returns (
+            uint256,
+            uint256,
+            uint256
+        )
+    {
+        return (sumOfLeftBalances, sumOfLeftBalances + balance, sumOfLeftBalances + balance + sumOfRightBalances);
+    }
+
+    /// @notice Add element to array
+    /// @param array Array to which the element must be added
+    /// @param toAdd Element to add
+    /// @return A new array with element added to it
+    function _addAddressToEncodedArray(bytes memory array, address toAdd) internal pure returns (bytes memory) {
+        address[] memory _currentNodePath = abi.decode(array, (address[]));
+        uint256 lengthOfNewPath = _currentNodePath.length + 1;
+
+        address[] memory _newNodePath = new address[](lengthOfNewPath);
+
+        for (uint256 index = 0; index < lengthOfNewPath - 1; index++) {
+            _newNodePath[index] = _currentNodePath[index];
+        }
+        _newNodePath[lengthOfNewPath - 1] = toAdd;
+        return abi.encode(_newNodePath);
+    }
+}
+
+// import "forge-std/console2.sol";
+
+abstract contract SelectorHelper is IClusterSelector {
+    /// @notice List of selected nodes
+    mapping(address => Node) public nodes;
+
+    /// @notice Total number of selected nodes in the tree
+    uint256 public totalElements;
+
+    /// @notice Address of the current root
+    address public root;
+
+    /// @notice Height of the tree at a given moment
+    /// @return Height of the tree
+    function heightOfTheTree() public view returns (uint256) {
+        return height(root);
+    }
+
+    /// @notice Height of any node at a given moment
+    /// @param node Address of the node whose height needs to be searched
+    /// @return Height of the node
+    function height(address node) public view returns (uint256) {
+        if (node == address(0)) return 0;
+        return nodes[node].height;
+    }
+
+    /// @notice Function to create a empty node
+    /// @param node Address of the new node
+    /// @param balance Balance of the new node
+    /// @return newNode Empty node with address and balance
+    function _newNode(address node, uint256 balance) internal pure returns (Node memory newNode) {
+        newNode = Node(node, balance, address(0), address(0), 0, 0, 1);
+    }
+
+    /// @notice Right rotate a given node
+    /// @param addressOfZ address of the node to right rotate
+    /// @return Returns the new root after the rotation
+    function _rightRotate(address addressOfZ) internal returns (address) {
+        if (addressOfZ == address(0)) {
+            revert(ClusterLib.CANNOT_RR_ADDRESS_ZERO);
+        }
+        Node storage z = nodes[addressOfZ];
+
+        Node storage y = nodes[z.left];
+
+        // do not rotate if left is 0
+        if (y.node == address(0)) {
+            // console2.log("RR: not because y is 0 ");
+            return z.node;
```

Comment:
shouldn't we revert in this case as right rotation should not occur in such cases as there is no left node to become the root.